man/octopi.1
.\"
.\"
.TH OCTOPI "1" "Septemper 2020"
.SH NAME
octopi - rule setting command for VxLAN extender daemon

.SH SYNOPSIS
.B octopi
[options...] command [parameters...]

.SH DESCRIPTION
.PP
This command defines and edit the rule database for octopi daemon\&.
.PP
See the documentation
.B octopi(8)
for ocotpi daemon\&.
.PP
.SH OPTIONS
.TP
.B \-h
show help and exit\&.
.TP
.B \-v
show version and exit\&.
.TP
.B \-c " path"
specify config file path\&. (default: /opt/octopi/etc/octopi.conf)
.TP
.B \-D " path"
specify the rule database path\&. (default: /var/octopi/octopidb)
.TP
.B \-Q " path"
specify rule sync queue path\&. (default: /var/octopi/octopiqueue)
.TP
.B \-S " number"
specify maximum number of parallel synchronization\&. (default: 8)
.TP
.B \-d
debug mode\&.
.TP
.B \-n
dry run mode\&.
.TP
.B \-q
quiet mode.
.TP
.B \-f
foreground mode\&. (output is line bufferd)\&.

.SH RULE COMMANDS
.TP
.B octopi [rule] init [pool_size  [hash_size]]
Initialize octopi rule database\&.
.TP
.B octopi [rule] list [vni [protocol]]
Show rule database as command list\&\&.
.TP
.B octopi [rule] show [vni [protocol]]
Show rule database structures as longest prefix match trees\&.
.TP
.B octopi [rule] find vni (address | protocol)
Lookup database and show relay destination\&.
If a protocol is specified instead of an address\&,
the broadcast destinations are displayed\&.
.TP
.B octopi [rule] add vni target relay [relay...]
Add new rule to the database\&.
.TP
.B octopi [rule] delete vni target [relay...]
Delete rule from the database\&.
If the relay is ommited, delete all relays\&.
.TP
.B octopi [rule] move vni target relay [relay...]
Change rule in database.
This command is same as delete then add, but do atomic timing\&.
.TP
.B octopi [rule] flush vni [proto]
Delete all trees from the database\&.
.TP
.B octopi [rule] save [vni [protocol]]
Write rules to standard output as save data format\&.

.RS 6
 rule vni
 target relay [relay...]
 ...
 (commit | abort)
.RS -6
.TP
.B octopi [rule] update [protocol]
Read save data from standard input and update the database\&.
.TP
.B octopi [rule] restore [protocol]
Read save data from standard input and replace the database\&.
This command is same as flush then update, but do atomic timing\&.

.SH SYNCHRONIZATION ACCESS CONTROL COMMANDS
.TP
.B octopi acl list [protocol]
Show ACL for synchronization parents as command list\&.
.TP
.B octopi acl add target action
Add ACL for synchronization parents\&.
.TP
.B octopi acl delete target action
Delete ACL for synchronization parents\&.

.SH SYNCHRONIZATION NODE COMMANDS
.TP
.B octopi backup list
Show backup node members as command list\&.
.TP
.B octopi backup add member [member ...]
Add backup node members nodes\&.
.TP
.B octopi backup delete [member ...]
Delete backup node members nodes\&.
If the member is ommited, delete all backup members\&.
.TP
.B octopi node list [relay ...]
Show synchronization nodes as command list\&.
.TP
.B octopi node add relay [member ...]
Add synchronization node members\&.
If the member is omitted, the same address as in relay is used\&.
.TP
.B octopi node delete relay [member ...]
Delete synchronization node members\&.
If the member is ommited, delete all members of relay\&.

.SH SYNCHRONIZATION QUEUE COMMANDS
.TP
.B octopi queue init [pool_size [hash_size]]
Initialize octopi synchronization queue\&.
.TP
.B octopi queue list [relay ...]
Show synchronization queue\&.
.TP
.B octopi queue show [relay ...]
Show synchronization nodes and queue structures\&.
.TP
.B octopi queue add [utype] node [vni]
Manually add synchronization schedules\&.
.TP
.B octopi queue delete [utype] node [vni]
Manually delete synchronization schedules\&.
.TP
.B octopi queue sync [[utype] node]
Same as \fBoctopi sync\fP, see below\&.

.SH ADMINISTRATION COMMANDS
.TP
.B octopi [rule] sync
Store rules on memory to the disk\&.
And start one-shot rule synchronization now\&.
If the node is ommited, synchronize all queued\&.
.TP
.B octopi [rule] pause
Set pause flags\&. The flag disable all rule temporarily\&.
.TP
.B octopi [rule] unpause
Unset pause flags\&. Enabling all rule again.
.TP
.B octopi [rule] fsck [normal | all]
Repair broken database file\&.
.TP
.B octopi [rule] dump [header | all]
Dump database for debugging\&.
.TP
.B octopi queue disable
Set queue disable flag, It disable synchronizing functions\&.
.TP
.B octopi queue enable
Unset queue disable flag. Enable synchronizing functions again\&.
.TP
.B octopi queue dump [header | all]
Dump synchronization queue for debugging\&.

.SH PARAMETERS
.TP
vni      :=  "all" | number | "any"
.br
specify Virtual Network Identifier (also known as VxLAN-ID).
\fB "all"\fP means all VNI rules\&. \fB "any"\fP means fallback rule\&.
.TP
address  :=  IPv4 Address | IPv6 Address
.TP
target   :=  address/mask | "default" | "default6"
.br
\fB "default"\fP match all IPv4, same as 0.0.0.0/0\&.
.br
\fB "default6"\fP match all IPv6, same as "::/0"\&.
.TP
relay    :=  addr | "drop" | "broadcast"
.br
specify relay target octopi daemon\&.
If the relay destination includes \fB "drop"\fP ,
the entire destination is disabled\&.
If the relay destination includes \fB "broadcast"\fP ,
relaying to broadcast destinations.

.TP
protocol :=  "ipv4" | "ipv6" | "all"
specify network protocol\&. \fB "all"\fP means both IPv4 and IPv6\&.

.TP
action   := "allow" | "deny"
.br
specify ACL action.

.TP
utype    := "auto" | "relay" | "member" | "both"
.br
specify node address type\&. If it is \fB "auto"\fP or ommited,
search for "relay" first\&. then if none, look for "member"\&.

.TP
node     := "all" | "backup" | address
.br
specify synchronization node address\&.
It is determined by the \fB utype\fP
to be a relay address or a member address\&.

.SH FILES
.PP
.B /opt/octopi/etc/octopi.conf
configuration file\&. See octopi.conf(8)\&.
.PP
/var/octopi/octopidb
.B octopi rule database\&.
.PP
.B /var/octopi/octopiqueue
octopi sync queue\&. See octopi(1)\&.

.SH AUTHOR
HANATAKA Shinya

.SH COPYRIGHT
Copyright \(co 2019 Internet Initiative Japan Inc\&.
.br
Copyright \(co 2019-2020 HANATAKA Shinya\&.

.SH "SEE ALSO"
octopid(8), octopi.conf(8)


src/octopi.c
/*
 *  octopi.c
 *
 *  copyright (c) 2019-2020 HANATAKA Shinya
 *  copyright (c) 2019-2020 Internet Initiative Japan Inc.
 */
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <pcap/pcap.h>

#include "octopi.h"
#include "setproctitle.h"
#include "command.h"
#include "logging.h"

/*
 *  global variables
 */
OD oddata;

/*
 *  finish snipper process (dummy)
 */
void
finish_snipper()
{
	crit_exit("Internal Error: dummy pcap_breakloop is called");
}

/*
 *  show usage
 */
void
usage_and_exit(int status)
{
	info("usage: %s [options] command [arguments...]", COMMAND_NAME);
	info("[options]");
	info("\t-h         show this help messages (and exit)");
	info("\t-v         show version (and exit)");
	info("\t-c path    specify config file");
	info("\t-D oath    specify rule file");
	info("\t-Q path    specify queue file");
	info("\t-S number  maximum number of parallel synchronization");
	info("\t-d         debug mode");
	info("\t-n         dryrun mode");
	info("\t-q         quiet mode");
	info("\t-f         line-buffered mode");
	info(" ");
	info("[command]");
	info("\trule, queue, node, backup, acl,");
	info("\tinit, show, list, find, add, delete, move, flush,");
	info("\tsave, restore, update, pause, unpause, sync, fsck, dump");
	info("\thelp, version");
	info(" ");
	info("More detailed usages can be found in the following way.");
	info("\tie) octopi add help");
	info("\tie) octopi acl add help");
	exit(status);
}

/*
 *  show version
 */
void
version_and_exit(int status)
{
	info("%s version %s", COMMAND_NAME, COMMAND_VERSION);
	exit(status);
}

void
command_help(OD *od, int ac, char *av[])
{
	usage_and_exit(EXIT_SUCCESS);
}

void
command_version(OD *od, int ac, char *av[])
{
	version_and_exit(EXIT_SUCCESS);
}

/*
 *  command list
 */
static struct command_list {
	char *name;
	void (*func)(OD*, int, char **);
} commands[] = {
	{ "init",    command_init,    },
	{ "show",    command_show,    },
	{ "list",    command_list,    },
	{ "find",    command_find,    },
	{ "add",     command_add,     },
	{ "delete",  command_delete,  },
	{ "move",    command_move,    },
	{ "flush",   command_flush,   },
	{ "pause",   command_pause,   },
	{ "unpause", command_unpause, },
	{ "save",    command_save,    },
	{ "restore", command_restore, },
	{ "update",  command_update,  },
	{ "sync",    command_sync,    },
	{ "fsck",    command_fsck,    },
	{ "dump",    command_dump,    },
	{ "queue",   command_queue,   },
	{ "node",    command_node,    },
	{ "backup",  command_backup,  },
	{ "acl",     command_acl,     },
	{ "rule",    command_rule,    },
	{ "help",    command_help,    },
	{ "version", command_version, },
	{ NULL,      NULL,            },
};

/*
 *  main routine
 */
int
main(int argc, char *argv[], char *envp[])
{
	OD *od = &oddata;
	int idx;
	struct command_list *p;

	/*
	 *  initialize process name
	 */
        setproctitle_init(argc, argv, envp);
	logging_init(COMMAND_NAME, LOG_USER);

	/*
	 *  initialize and read config
	 */
	initialize(od);
	idx = parse_args(od, argc, argv);
	readconf(od);
	od->proc_type = PROC_TYPE_COMMAND;
	logging_init(COMMAND_NAME, od->log_facility);

	/*
	 *  set standard I/O line bufferd
	 */
	if (od->foreground) {
		setvbuf(stdin,  NULL, _IOLBF, 0);
		setvbuf(stdout, NULL, _IOLBF, 0);
		setvbuf(stderr, NULL, _IOLBF, 0);
	}

	/*
	 *  setup rule and queue file
	 */
	if (od->opt_rule_file)
		od->rule_file = od->opt_rule_file;
	if (od->opt_queue_file)
		od->queue_file = od->opt_queue_file;

	/*
	 *  execute commands
	 */
	if (idx >= argc)
		usage_and_exit(EXIT_FAILURE);

	for (p = commands; p->name; ++ p) {
		if (strcasecmp(p->name, argv[idx]) == 0) {
			p->func(od, argc - idx - 1, argv + idx + 1);
			exit(EXIT_SUCCESS);
		}
	}

	error("%s: unknown command", COMMAND_NAME);
	usage_and_exit(EXIT_FAILURE);

	/* NOTREACHED */
	return EXIT_FAILURE;
}
